### PR TITLE
[msal-common] Add knownAuthorities parameter to support B2C validation

### PR DESCRIPTION
**Pull request by sameerag**
This PR adds B2C support for `msal-common`; This also ports the changes needed for knownAuthorities() support from #1276.

cc @tnorling

### CODE REVIEW

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts [hunk 1]
```diff
@@ -20,14 +31,14 @@ export class AuthorityFactory {
...
-        switch (pathSegments[0]) {
-            case "tfp":
-                // tfp denotes a b2c url
-                return AuthorityType.B2C;
-            default:
-                // default authority is always AAD
-                return AuthorityType.Aad;
-        }
+
+        if (pathSegments[0] === "adfs")
```
Comment (sangonzal) [inline lib/msal-common/src/auth/authority/AuthorityFactory.ts:35]: Should this be case insensitive comparison.

File: lib/msal-common/src/auth/authority/B2cAuthority.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/*
```
Comment (sangonzal) [inline lib/msal-common/src/auth/authority/B2cAuthority.ts:1]: Comments in whole file should be updated from AAD to B2C.

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts (hunk 1 above)
Comment (tnorling) [reply lib/msal-common/src/auth/authority/AuthorityFactory.ts:35]: Yes, good catch. Will update

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts [hunk 3]
```diff
@@ -4,14 +4,25 @@
...
 import { UrlString } from "../../url/UrlString";
 
-export class AuthorityFactory {
+export class AuthorityFactory {  
+    /**
+     * Use when Authority is B2C to provide list of trusted/allowed domains.
+     */
+    public static setKnownAuthorities(knownAuthorities: Array<string>): void {
+        if (!B2CTrustedHostList.length){
+            knownAuthorities.forEach(function(authority){
```
Comment (DarylThayil) [inline lib/msal-common/src/auth/authority/AuthorityFactory.ts:21]: reason not to just use an arrow function here ?

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts [hunk 4]
```diff
@@ -20,14 +31,14 @@ export class AuthorityFactory {
...
-        switch (pathSegments[0]) {
-            case "tfp":
-                // tfp denotes a b2c url
-                return AuthorityType.B2C;
-            default:
-                // default authority is always AAD
-                return AuthorityType.Aad;
-        }
+
+        if (pathSegments[0].toLowerCase() === "adfs")
```
Comment (DarylThayil) [inline lib/msal-common/src/auth/authority/AuthorityFactory.ts:35]: is this a constant for adfs?

File: lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts [hunk 5]
```diff
@@ -14,26 +16,58 @@ describe("AuthorityFactory.ts Class Unit Tests", () => {
             return null;
         }
     };
+
+    beforeEach(() => {
```
Comment (DarylThayil) [inline lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts:20]: what does this do for the test?

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts (hunk 4 above)
Comment (tnorling) [reply lib/msal-common/src/auth/authority/AuthorityFactory.ts:35]: Yes, confirmed this. But this is also mostly a placeholder until we support adfs

File: lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts (hunk 5 above)
Comment (tnorling) [reply lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts:20]: Cleans up the static host list between tests so the B2C tests dont break the AAD tests.

File: lib/msal-common/src/auth/authority/B2cAuthority.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { Authority } from "./Authority";
+import { AuthorityType } from "./AuthorityType";
+import { ClientConfigurationError } from "../../error/ClientConfigurationError";
+import { INetworkModule } from "../../network/INetworkModule";
+
+export const B2CTrustedHostList: string[] = [];
```
Comment (jasonnutter) [inline lib/msal-common/src/auth/authority/B2cAuthority.ts:10]: Could this instead be a static member on the b2cauthority class?

File: lib/msal-common/src/auth/authority/B2cAuthority.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+        throw ClientConfigurationError.createUntrustedAuthorityError();
+    }
+
+    /**
+     * Checks to see if the host is in a list of trusted hosts
+     * @param {string} The host to look up
+     */
+    private isInTrustedHostList(host: string): boolean {
+        return B2CTrustedHostList.includes(host);
```
Comment (jasonnutter) [inline lib/msal-common/src/auth/authority/B2cAuthority.ts:42]: `.includes` doesn't work on IE.

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts [hunk 8]
```diff
@@ -4,14 +4,25 @@
...
 
-export class AuthorityFactory {
+export class AuthorityFactory {  
+    /**
+     * Use when Authority is B2C to provide list of trusted/allowed domains.
+     */
+    public static setKnownAuthorities(knownAuthorities: Array<string>): void {
+        if (!B2CTrustedHostList.length){
+            knownAuthorities.forEach(function(authority){
+                B2CTrustedHostList.push(authority);
```
Comment (jasonnutter) [inline lib/msal-common/src/auth/authority/AuthorityFactory.ts:22]: I would personally prefer some sort of method on the b2cauthority class for adding an authority. This is leaking implementation details.

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts (hunk 4 above)
Comment (jasonnutter) [reply lib/msal-common/src/auth/authority/AuthorityFactory.ts:35]: Can this string be added to a constants file?

File: lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts [hunk 9]
```diff
@@ -14,26 +16,58 @@ describe("AuthorityFactory.ts Class Unit Tests", () => {
...
+        while(B2CTrustedHostList.length) {
+            B2CTrustedHostList.pop();
```
Comment (jasonnutter) [inline lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts:22]: This also leaks implementation details.

File: lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts (hunk 5 above)
Comment (jasonnutter) [reply lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts:20]: A commend would probably helpful here to explain that.

File: lib/msal-common/src/auth/authority/B2cAuthority.ts (hunk 7 above)
Comment (tnorling) [reply lib/msal-common/src/auth/authority/B2cAuthority.ts:42]: Yikes. This is how it's done in core as of 1.3...

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts (hunk 4 above)
Comment (tnorling) [reply lib/msal-common/src/auth/authority/AuthorityFactory.ts]: Done

File: lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts (hunk 5 above)
Comment (tnorling) [reply lib/msal-common/test/auth/authority/AuthorityFactory.spec.ts:20]: Done

File: lib/msal-common/src/auth/authority/B2cAuthority.ts (hunk 6 above)
Comment (tnorling) [reply lib/msal-common/src/auth/authority/B2cAuthority.ts]: Done

File: lib/msal-common/src/auth/authority/B2cAuthority.ts (hunk 7 above)
Comment (tnorling) [reply lib/msal-common/src/auth/authority/B2cAuthority.ts]: Updated to `.indexOf`

File: lib/msal-common/src/auth/authority/AuthorityFactory.ts (hunk 8 above)
Comment (tnorling) [reply lib/msal-common/src/auth/authority/AuthorityFactory.ts]: Moved to B2CAuthority class

**Pull request merged**